### PR TITLE
fix: keep events order

### PR DESCRIPTION
**Pull request by theomonnom**
- Always ensure we keep the order from the FfiServer
   - So the complex logic can be handled from Rust (for synchronisation)
   - E.g: the ffi server ensure that the `LocalTrackPublished` callback is sent before the `LocalTrackPublished` event
- We know use only one FfiQueue per room (So it is possible for us to keep things synchronised)

The room musts be used inside one event_loop (specified inside the ctor)
AudioStream & VideoStream can be used on other event_loops (making it easy to have another IO thread for e.g)

### CODE REVIEW

File: livekit/_utils.py [hunk 1]
```diff
@@ -22,3 +24,85 @@ async def get(self) -> T:
             await self._event.wait()
         self._event.clear()
         return self._queue.popleft()
+
+
+class BroadcastQueue(Generic[T]):
```
Comment (theomonnom) [inline livekit/_utils.py:29]: Better naming for BroadcastQueue & BroadcastChannel.
Feels confusing rn

File: livekit/_utils.py [hunk 2]
```diff
@@ -22,3 +24,90 @@ async def get(self) -> T:
...
         self._event.clear()
         return self._queue.popleft()
+
+
+class Queue(asyncio.Queue[T]):
+    def __init__(self,
+                 loop: Optional[asyncio.AbstractEventLoop] = None,
+                 maxsize: int = 0) -> None:
+        super().__init__(maxsize)
+        self._loop = loop or asyncio.get_event_loop()
```
Comment (theomonnom) [inline livekit/_utils.py:34]: Am I right to assume self._loop will be used for the asyncio.Queue?

File: livekit/_utils.py (hunk 1 above)
Comment (davidzhao) [reply livekit/_utils.py:37]: agreed it's a bit confusing.

also a bit confusing that `subscribe` returns a queue for clients to manage.

Comment (davidzhao) [reply livekit/_utils.py:37]: if the main difference between _Queue and _Channel is thread safety? why not call it ThreadsafeBroadcastQueue?

File: livekit/_utils.py (hunk 2 above)
Comment (davidzhao) [reply livekit/_utils.py:34]: how would it know to use self._loop?

Comment (theomonnom) [reply livekit/_utils.py:34]: this class inherits asyncio.Queue, which also use self._loop.

Comment (davidzhao) [reply livekit/_utils.py:34]: ok, I don't think it's safe to override loop in that way.. 

https://github.com/python/cpython/blob/b88d9e75f68f102aca45fa62e2b0e2e2ff46d810/Lib/asyncio/mixins.py#L9

Comment (theomonnom) [reply livekit/_utils.py:34]: right

Comment (theomonnom) [reply livekit/_utils.py:34]: it is only being overrided when not used but I guess I could try to find a better solution

**Comment by theomonnom** [comment]
it's tricky, now with the changes, we can't publish tracks before running room.run() ...

File: examples/publish_hue.py [hunk 3]
```diff
@@ -58,32 +58,24 @@ async def main():
...
 
     loop = asyncio.get_event_loop()
-    main_task = asyncio.ensure_future(main())
+    asyncio.ensure_future(main())
     for signal in [SIGINT, SIGTERM]:
-        loop.add_signal_handler(signal, main_task.cancel)
+        loop.add_signal_handler(signal, loop.stop)
     try:
-        loop.run_until_complete(main_task)
+        loop.run_forever()
```
Comment (davidzhao) [inline examples/publish_hue.py:79]: nice!

File: examples/basic_room.py [hunk 4]
```diff
@@ -6,7 +6,7 @@
 import livekit
 
 URL = 'ws://localhost:7880'
-TOKEN = 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJleHAiOjE5MDY2MTMyODgsImlzcyI6IkFQSVRzRWZpZFpqclFvWSIsIm5hbWUiOiJuYXRpdmUiLCJuYmYiOjE2NzI2MTMyODgsInN1YiI6Im5hdGl2ZSIsInZpZGVvIjp7InJvb20iOiJ0ZXN0Iiwicm9vbUFkbWluIjp0cnVlLCJyb29tQ3JlYXRlIjp0cnVlLCJyb29tSm9pbiI6dHJ1ZSwicm9vbUxpc3QiOnRydWV9fQ.uSNIangMRu8jZD5mnRYoCHjcsQWCrJXgHCs0aNIgBFY'
```
Comment (davidzhao) [inline examples/basic_room.py:9]: can't wait till we have server-sdk integrated so this isn't necessary

File: livekit/room.py [hunk 5]
```diff
@@ -64,6 +70,7 @@ def name(self) -> str:
     def metadata(self) -> str:
         return self._info.metadata
 
+    @property
     def e2ee_manager(self) -> E2EEManager:
```
Comment (davidzhao) [inline livekit/room.py:74]: would this change the interface? let's document in changelog

**Review by davidzhao — APPROVED** [review]
lgtm!

File: livekit/room.py (hunk 5 above)
Comment (theomonnom) [reply livekit/room.py:74]: 👍

**Pull request merged**